#include "mainwindow.h"
#include "ui_mainwindow.h"
#include "autodetect_manager.h"
#include "serialport_manager.h"
#include <QDateTime>
#include <QMessageBox>
#include <QScrollBar>
#include<QDebug>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , m_autoDetect(new AutoDetectManager(this))
    , m_serialManager(new SerialPortManager(this))
    , m_isConnected(false)
{
    ui->setupUi(this);
    setupUI();
    setupConnections();

    // å¯åŠ¨è‡ªåŠ¨æ£€æµ‹
    m_autoDetect->startDetection(1500);
    refreshPortList();

    logMessage("ServeDebug è‡ªåŠ¨æ£€æµ‹ç³»ç»Ÿå·²å¯åŠ¨");
}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::setupUI()
{
    setWindowTitle("ServeDebug AutoDetect v2.0");
    setMinimumSize(800, 600);

    // è®¾ç½®çŠ¶æ€æ 
    ui->statusbar->showMessage("å°±ç»ª");

    // åˆå§‹åŒ–æ§ä»¶çŠ¶æ€
    updateConnectionStatus(false);
}

void MainWindow::setupConnections()
{
    // è‡ªåŠ¨æ£€æµ‹ä¿¡å·
    connect(m_autoDetect, &AutoDetectManager::debuggerDetected,
            this, &MainWindow::onDebuggerDetected);
    connect(m_autoDetect, &AutoDetectManager::debuggerRemoved,
            this, &MainWindow::onDebuggerRemoved);

    // ä¸²å£ç®¡ç†ä¿¡å·
    connect(m_serialManager, &SerialPortManager::connected,
            this, &MainWindow::onSerialConnected);
    connect(m_serialManager, &SerialPortManager::disconnected,
            this, &MainWindow::onSerialDisconnected);
    connect(m_serialManager, &SerialPortManager::errorOccurred,
            this, [this](const QString &error) { logMessage(error, true); });

    // æ„é€ å‡½æ•°
    connect(ui->btnClearLog, &QPushButton::clicked,ui->textLog, &QPlainTextEdit::clear);
}

void MainWindow::onDebuggerDetected(const QString &portName, const QString &description)
{
    QString message = QString("ğŸ” æ£€æµ‹åˆ° ServeDebug è®¾å¤‡: %1 (%2)").arg(portName).arg(description);
    logMessage(message);
    // è‡ªåŠ¨é€‰æ‹©æ–°æ£€æµ‹åˆ°çš„è®¾å¤‡
    for (int i = 0; i < ui->listDevices->count(); ++i) {
        QListWidgetItem *item = ui->listDevices->item(i);
        if (item->text().contains(portName)) {
            ui->listDevices->setCurrentItem(item);
            break;
        }
    }

    // å¯é€‰ï¼šè‡ªåŠ¨è¿æ¥ï¼ˆå–æ¶ˆæ³¨é‡Šå¯ç”¨ï¼‰
     if (!m_isConnected) {
      m_serialManager->connectToPort(portName);
     }
}

void MainWindow::onDebuggerRemoved(const QString &portName)
{
    QString message = QString("âŒ è®¾å¤‡å·²ç§»é™¤: %1").arg(portName);
    logMessage(message);

    // å¦‚æœå½“å‰è¿æ¥çš„è®¾å¤‡è¢«ç§»é™¤ï¼Œè‡ªåŠ¨æ–­å¼€
    if (m_isConnected && m_serialManager->currentPort() == portName) {
        m_serialManager->disconnectPort();
    }
}

void MainWindow::onSerialConnected(const QString &portName)
{
    m_isConnected = true;
    updateConnectionStatus(true);

    QString message = QString("âœ… å·²è¿æ¥åˆ°: %1").arg(portName);
    logMessage(message);
    ui->labelCurrentPort->setText(QStringLiteral("ç«¯å£: %1").arg(portName));
    ui->statusbar->showMessage(message);
}

void MainWindow::onSerialDisconnected()
{
    m_isConnected = false;
    updateConnectionStatus(false);

    logMessage("ğŸ”Œ è¿æ¥å·²æ–­å¼€");
    ui->statusbar->showMessage("è¿æ¥å·²æ–­å¼€");
}

void MainWindow::on_btnManualRefresh_clicked()
{
    refreshPortList();
    logMessage("æ‰‹åŠ¨åˆ·æ–°è®¾å¤‡åˆ—è¡¨");
}

void MainWindow::on_btnConnect_clicked()
{
    QListWidgetItem *currentItem = ui->listDevices->currentItem();
    if (!currentItem) {
        QMessageBox::warning(this, "è­¦å‘Š", "è¯·å…ˆé€‰æ‹©ä¸€ä¸ªè®¾å¤‡");
        return;
    }

    QString portName = currentItem->text().split(" - ").first();
    logMessage(QString("æ­£åœ¨è¿æ¥: %1").arg(portName));
    m_serialManager->connectToPort(portName);

}

void MainWindow::on_btnDisconnect_clicked()
{
    m_serialManager->disconnectPort();
}

void MainWindow::refreshPortList()
{
    ui->listDevices->clear();

    QList<QSerialPortInfo> ports = QSerialPortInfo::availablePorts();
    for (const QSerialPortInfo &port : ports) {
        QString itemText = QString("%1 - %2")
            .arg(port.portName())
            .arg(port.description());
        ui->listDevices->addItem(itemText);
    }

    ui->labelDeviceCount->setText(QString("æ‰¾åˆ° %1 ä¸ªè®¾å¤‡").arg(ports.count()));
}

void MainWindow::logMessage(const QString &message, bool isError)
{
    QString timestamp = QDateTime::currentDateTime().toString("hh:mm:ss");
    QString logEntry = QString("[%1] %2").arg(timestamp).arg(message);

    QTextCharFormat format;
    if (isError) {
        format.setForeground(QBrush(Qt::red));
    } else {
        format.setForeground(QBrush(Qt::blue));
    }

    QTextCursor cursor(ui->textLog->document());
    cursor.movePosition(QTextCursor::End);
    cursor.insertText(logEntry + "\n", format);

    // è‡ªåŠ¨æ»šåŠ¨
    QScrollBar *scrollbar = ui->textLog->verticalScrollBar();
    scrollbar->setValue(scrollbar->maximum());
}

void MainWindow::updateConnectionStatus(bool connected)
{
    ui->btnConnect->setEnabled(!connected);
    ui->btnDisconnect->setEnabled(connected);

    if (connected) {
        ui->labelStatus->setText("ğŸŸ¢ å·²è¿æ¥");
        ui->labelStatus->setStyleSheet("color: green; font-weight: bold;");
    } else {
        ui->labelStatus->setText("ğŸ”´ æœªè¿æ¥");
        ui->labelStatus->setStyleSheet("color: red;");
    }
}

